kafka-connector/src/main/java/com/google/pubsub/kafka/source/CloudPubSubSourceTask.java

```diff
@@ -197,21 +197,28 @@ private void ackMessages() {
     if (ackIds.size() != 0) {
       AcknowledgeRequest.Builder requestBuilder = AcknowledgeRequest.newBuilder()
           .setSubscription(cpsSubscription);
+      final Set<String> ackIdsCopy = new HashSet<>();
       synchronized (ackIds) {
         requestBuilder.addAllAckIds(ackIds);
-        ackIds.clear();
+        ackIdsCopy.addAll(ackIds);
```

Comment:
It seems like this could result in repeatedly sending acks for the same messages. While a call to ackMessages is outstanding, if another poll occurs, then all acks will be sent again. Perhaps it would be better to keep a "acksInFlight" HashSet that we move the items from ackIds to. Then, in poll, we exclude messages that are in ackIds or in acksInFlight. Once the Future returns here, we remove the items from acksInFlight.

Reply:
Thanks for the feedback. I was aware of the trade off ( sending repeated acks ) but by looking at the `SubscriberGrpc` java docs, repeated acks should not result in errors on Google side. Anyways, having global variable would eliminate repeated acks so lets go for that